### PR TITLE
Fix #3279: Created BUILD.bazel in feedbackreporting inside domain module

### LINKED ISSUE #3279: Create a BUILD.bazel file for domain/feedbackreporting/
This task is part of the build system migration from Gradle to Bazel. 

## Prerequesites

- Make sure you are able to build Oppia using Bazel
- Follow the [Oppia Bazel Setup instructions](https://github.com/oppia/oppia-android/wiki/Oppia-Bazel-Setup-Instructions) to setup Bazel and learn the terminology.
- Read the [Gradle Bazel Migration Best Practies and FAQ](https://github.com/oppia/oppia-android/wiki/Gradle--Bazel-Migration-Best-Practices-and-FAQ) for more context and see if there's an answer there to any question you have.

## Task

1. Create a BUILD.bazel file in

`domain/src/main/java/org/oppia/android/domain/feedbackreporting`

1. Define the kt_android_library library/libraries

Based on the imports in the Kotlin files, add the required third_party dependencies to make the libraries build. Check the Useful links section for a list of third_party dependencies.  You may also need to add the non third_party dependencies.

This package should be split into two libraries, one for the annotation and the other for the module.

3. Add the directory to the MIGRATED_PROD_FILES of the module's root BUILD.bazel file

4. Add the libraries to the list of deps dependencies in the targets that require the library

You will have to set the apropriate visbility. See the list of visibilities in the link below.

5. Clean up any unused deps dependencies

6. Verify that the app builds

## Useful links

- [BUILD.bazel files](https://github.com/oppia/oppia-android/search?q=filename%3ABUILD.bazel&type=code)
- [Visibilities](https://github.com/oppia/oppia-android/blob/76dc13dba3ad77aecd2218a5372949d6551f12fe/BUILD.bazel)
- [third_party dependencies](https://github.com/oppia/oppia-android/blob/a70d9b016f0d77dd30c4939041f82ea02ee28d1d/third_party/versions.bzl)

### PR DESCRIPTION
**Pull request by deepanshu731**
Fixed #3279 
Created BUILD.bazel in feedbackreporting

## Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

### CODE REVIEW

File: domain/src/main/java/org/oppia/android/domain/feedbackreporting/BUILD.bazel [hunk 1]
```diff
@@ -0,0 +1,31 @@
+"""
+Utilities corresponding to feedbackreporting.
```
Comment (prayutsu) [inline domain/src/main/java/org/oppia/android/domain/feedbackreporting/BUILD.bazel]:
```suggestion
Utilities corresponding to feedback reporting.
```

File: domain/src/main/java/org/oppia/android/domain/feedbackreporting/BUILD.bazel [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+kt_android_library(
+    name = "report_schema_version",
+    srcs = [
+        "ReportSchemaVersion.kt",
+    ],
+    visibility = ["//:oppia_api_visibility"],
+    deps = [
+        "//third_party:javax_inject_javax_inject",
+    ],
+)
```
Comment (prayutsu) [inline domain/src/main/java/org/oppia/android/domain/feedbackreporting/BUILD.bazel]: Why not combine the two libraries into one?

**Review by prayutsu — CHANGES_REQUESTED** [review]
@deepanshu731 Left some comments, PTAL.
Also please update the PR description, all points in checkboxes should be followed and then marked as checked.

File: domain/src/main/java/org/oppia/android/domain/feedbackreporting/BUILD.bazel (hunk 2 above)
Comment (deepanshu731) [reply domain/src/main/java/org/oppia/android/domain/feedbackreporting/BUILD.bazel]: done

**Comment by prayutsu** [comment]
@deepanshu731 You shouldn't mark conversations resolved that are not initiated by you. The conversation threads should only be resolved by the person who started it. They help the reviewers to keep track of the changes that aren't verified by them.

**Review by prayutsu — APPROVED** [review]
LGTM, Thanks @deepanshu731

**Comment by deepanshu731** [comment]
> @deepanshu731 You shouldn't mark conversations resolved that are not initiated by you. The conversation threads should only be resolved by the person who started it. They help the reviewers to keep track of the changes that aren't verified by them.

Actually, I didn't resolve that when I committed the changes requested by you, it automatically showed  resolved

**Comment by prayutsu** [comment]
> Actually, I didn't resolve that when I committed the changes requested by you, it automatically showed resolved

I unresolved the already resolved conversations, and mark them resolved again after verifying them. You may have marked it resolved accidentally (It looks like you pressed resolve conversation instead of pressing the `comment` button).

File: domain/src/main/java/org/oppia/android/domain/feedbackreporting/BUILD.bazel [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+"""
+
+load("@dagger//:workspace_defs.bzl", "dagger_rules")
+load("@io_bazel_rules_kotlin//kotlin:kotlin.bzl", "kt_android_library")
+
+kt_android_library(
+    name = "feedback_reporting_module",
+    srcs = [
+        "FeedbackReportingModule.kt",
+        "ReportSchemaVersion.kt",
```
Comment (BenHenning) [inline domain/src/main/java/org/oppia/android/domain/feedbackreporting/BUILD.bazel]: Conventionally, we keep modules separate by putting them in a library called "prod_module" with module visibility (rather than API visibility). I suggest splitting this apart in that way. ReportSchemaVersion can go in a separate library called report_schema_version (it's generally nice to isolate annotations since they are generally used individually).

File: domain/src/main/java/org/oppia/android/domain/feedbackreporting/BUILD.bazel [hunk 4]
```diff
@@ -0,0 +1,21 @@
+"""
+Utilities corresponding to feedback reporting.
```
Comment (BenHenning) [inline domain/src/main/java/org/oppia/android/domain/feedbackreporting/BUILD.bazel]: These aren't quite utilities. Suggest instead:

``Domain services & definitions corresponding to user feedback reporting.``

to be a bit more specific.

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @deepanshu731! Had a couple of comments.

@jcqli you may also want to review this since you're actively making a lot of changes to these directories.

File: domain/src/main/java/org/oppia/android/domain/feedbackreporting/BUILD.bazel (hunk 4 above)
Comment (deepanshu731) [reply domain/src/main/java/org/oppia/android/domain/feedbackreporting/BUILD.bazel]: done

File: domain/src/main/java/org/oppia/android/domain/feedbackreporting/BUILD.bazel (hunk 3 above)
Comment (deepanshu731) [reply domain/src/main/java/org/oppia/android/domain/feedbackreporting/BUILD.bazel]: done

File: domain/src/main/java/org/oppia/android/domain/feedbackreporting/BUILD.bazel [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+        ":report_schema_version",
+    ],
+)
+
+kt_android_library(
+    name = "report_schema_version",
+    srcs = [
+        "ReportSchemaVersion.kt",
+    ],
+    visibility = ["//:oppia_prod_module_visibility"],
```
Comment (jcqli) [inline domain/src/main/java/org/oppia/android/domain/feedbackreporting/BUILD.bazel]: Since this is a dagger constant should this be made private? @BenHenning

**Comment by jcqli** [comment]
> @jcqli you may also want to review this since you're actively making a lot of changes to these directories.

@BenHenning overall makes sense to me -- should the test files also be migrated in this PR?

File: domain/src/main/java/org/oppia/android/domain/feedbackreporting/BUILD.bazel (hunk 5 above)
Comment (BenHenning) [reply domain/src/main/java/org/oppia/android/domain/feedbackreporting/BUILD.bazel]: Is it only meant to be used by the feedbackreporting package @jcqli? If so, then yes. If not, then it should be API visibility since it provides the qualifier to inject whatever property it's corresponding to.

File: .idea/runConfigurations.xml [hunk 6]
```diff
@@ -1,12 +0,0 @@
-<?xml version="1.0" encoding="UTF-8"?>
```
Comment (BenHenning) [inline .idea/runConfigurations.xml]: Please don't remove unrelated files to your PR.

Could you revert this?

**Review by BenHenning — COMMENTED** [review]
Thanks @deepanshu731! Couple of follow-up comments yet.

File: domain/src/main/java/org/oppia/android/domain/feedbackreporting/BUILD.bazel (hunk 5 above)
Comment (jcqli) [reply domain/src/main/java/org/oppia/android/domain/feedbackreporting/BUILD.bazel]: I imagine it'll only be used when constructing the report request being sent to the backend so I think this should be private.

File: .idea/runConfigurations.xml (hunk 6 above)
Comment (deepanshu731) [reply .idea/runConfigurations.xml]: sorry by mistake, done

File: domain/src/main/java/org/oppia/android/domain/feedbackreporting/BUILD.bazel (hunk 5 above)
Comment (deepanshu731) [reply domain/src/main/java/org/oppia/android/domain/feedbackreporting/BUILD.bazel]: what do I have to do next @BenHenning

Comment (jcqli) [reply domain/src/main/java/org/oppia/android/domain/feedbackreporting/BUILD.bazel]: I believe if you don't specify visibility at all (remove this line) then it defaults to private.

Comment (deepanshu731) [reply domain/src/main/java/org/oppia/android/domain/feedbackreporting/BUILD.bazel]: if i didnt specify the visibility, its giving me this error
Screenshot from 2021-08-20 14-48-52
Instead if i give visibility like( visibility = ["//domain:__pkg__"],) it removes errors and its only visible to domain pkg, Is this correct @jcqli

Comment (BenHenning) [reply domain/src/main/java/org/oppia/android/domain/feedbackreporting/BUILD.bazel]: @deepanshu731 I think that's because you added ``"//domain/src/main/java/org/oppia/android/domain/feedbackreporting:report_schema_version",`` to domain/BUILD.bazel. Does it need to be there/does the app build without that dep listed? If so, this can be private. If not, this should be marked as API visibility.

File: domain/src/main/java/org/oppia/android/domain/feedbackreporting/BUILD.bazel [hunk 7]
```diff
@@ -0,0 +1,31 @@
+"""
+Domain services & definitions corresponding to user feedback reporting.
+"""
+
+load("@dagger//:workspace_defs.bzl", "dagger_rules")
+load("@io_bazel_rules_kotlin//kotlin:kotlin.bzl", "kt_android_library")
+
+kt_android_library(
+    name = "feedback_reporting_module",
```
Comment (BenHenning) [inline domain/src/main/java/org/oppia/android/domain/feedbackreporting/BUILD.bazel]: Please call this prod_module for consistency.

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @deepanshu731. Just 2 follow-ups!

File: domain/src/main/java/org/oppia/android/domain/feedbackreporting/BUILD.bazel (hunk 7 above)
Comment (deepanshu731) [reply domain/src/main/java/org/oppia/android/domain/feedbackreporting/BUILD.bazel]: done

File: domain/src/main/java/org/oppia/android/domain/feedbackreporting/BUILD.bazel (hunk 5 above)
Comment (deepanshu731) [reply domain/src/main/java/org/oppia/android/domain/feedbackreporting/BUILD.bazel]: thanks @BenHenning , it can be made private

**Comment by BenHenning** [comment]
@deepanshu731 ReportSchemaVersionTest is now failing in the CI run since domain's TEST_DEPS aren't including these dependencies. The latest change is resulting in the files not being included anywhere in the build--I don't think that's right. In the very least, they seem needed for TEST_DEPS.

@jcqli is it correct that the module or schema version isn't yet referenced since downstream changes haven't landed yet?

**Comment by BenHenning** [comment]
@deepanshu731 ReportSchemaVersionTest is now failing in the CI run since domain's TEST_DEPS aren't including these dependencies. The latest change is resulting in the files not being included anywhere in the build--I don't think that's right. In the very least, they seem needed for TEST_DEPS.

@jcqli is it correct that the module or schema version isn't yet referenced since downstream changes haven't landed yet?

**Comment by jcqli** [comment]
> @jcqli is it correct that the module or schema version isn't yet referenced since downstream changes haven't landed yet?

This is correct -- they are not referenced yet. I believe they should be included in the TEST_DEPS, though, since there are [tests for the Report Schema Version](https://github.com/oppia/oppia-android/blob/develop/domain/src/test/java/org/oppia/android/domain/feedbackreporting/ReportSchemaVersionTest.kt). Does that sound right?

**Comment by deepanshu731** [comment]
@BenHenning i have included report_schema_version in TEST_DEPS but reportschemaversiontest is still failing.Can you help me out
Screenshot from 2021-08-24 23-38-51
this is the error log
I think some more dependencies have to be added in report_schema_version

File: domain/BUILD.bazel [hunk 8]
```diff
@@ -123,6 +124,7 @@ TEST_DEPS = [
     ":testing",
     "//data/src/main/java/org/oppia/android/data/backends/gae/model",
     "//data/src/main/java/org/oppia/android/data/persistence:cache_store",
+    "//domain/src/main/java/org/oppia/android/domain/feedbackreporting:report_schema_version",
```
Comment (BenHenning) [inline domain/BUILD.bazel:128]: The module also needs to be included as a dep @deepanshu731. I think that's why you're getting an error.

You can verify this is correct locally by trying to run the test that's failing in CI:

```sh
bazel test //domain:src/test/java/org/oppia/android/domain/feedbackreporting/ReportSchemaVersionTest
```

If that builds & passes locally, you've got the deps correct.

File: domain/src/main/java/org/oppia/android/domain/feedbackreporting/BUILD.bazel [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+
+load("@dagger//:workspace_defs.bzl", "dagger_rules")
+load("@io_bazel_rules_kotlin//kotlin:kotlin.bzl", "kt_android_library")
+
+kt_android_library(
+    name = "prod_module",
+    srcs = [
+        "FeedbackReportingModule.kt",
+    ],
+    deps = [
```
Comment (BenHenning) [inline domain/src/main/java/org/oppia/android/domain/feedbackreporting/BUILD.bazel:14]: I think you want to keep prod module visibility here; it will eventually be used externally.

**Review by BenHenning — COMMENTED** [review]
Thanks @deepanshu731. Just 2 comments, then I think the PR can be merged if CI is passing.

Yep that helps @jcqli. I think we'll want to keep the annotation as test-only for now, but we can change that in subsequent PRs if broader visibility is needed.

File: domain/src/main/java/org/oppia/android/domain/feedbackreporting/BUILD.bazel (hunk 9 above)
Comment (deepanshu731) [reply domain/src/main/java/org/oppia/android/domain/feedbackreporting/BUILD.bazel:14]: done

File: domain/BUILD.bazel (hunk 8 above)
Comment (deepanshu731) [reply domain/BUILD.bazel:128]: thanks @BenHenning , test passed locally

**Comment by BenHenning** [comment]
Hi @deepanshu731. My apologies for the delay--I'll need to take a look at this tomorrow.

**Comment by BenHenning** [comment]
Hi--sorry again about the delay. I still plan to look at this later in the week; I've just had some project emergencies to deal with in the meantime. Thanks for your patience.

**Review by BenHenning — APPROVED** [review]
Thanks @deepanshu731. So sorry for the delay here. LGTM.
